tests/cron.test.ts

```diff
@@ -1237,7 +1148,152 @@ describe('cron', () => {
 		);
 		clock.tick(1500);
 		job.stop();
-		clock.restore();
 		expect(callback).toHaveBeenCalledTimes(1);
 	});
+
+	describe('waitForCompletion and job status tracking', () => {
+		it('should wait for async job completion when waitForCompletion is true', async () => {
+			const clock = sinon.useFakeTimers();
+			let isJobCompleted = false;
+
+			const job = new CronJob(
+				'*/2 * * * * *',
+				async () => {
+					expect(job.isCallbackRunning).toBe(true);
+					await new Promise<void>(resolve => {
+						setTimeout(() => {
+							callback();
+							isJobCompleted = true;
+							resolve();
+						}, 1000);
+					});
+				},
+				null,
+				true,
+				null,
+				null,
+				false,
+				null,
+				false,
+				true // waitForCompletion: true
+			);
+
+			expect(job.isCallbackRunning).toBe(false);
+
+			// First execution
+			await clock.tickAsync(2000);
+			expect(job.isCallbackRunning).toBe(true);
+
+			// Wait for job completion
+			await clock.tickAsync(500);
+			expect(isJobCompleted).toBe(false);
+			expect(job.isCallbackRunning).toBe(true);
+
+			await clock.tickAsync(1000);
+			expect(isJobCompleted).toBe(true);
+			expect(job.isCallbackRunning).toBe(false);
+
+			job.stop();
+		});
+
+		it('should not wait for job completion when waitForCompletion is false', () => {
+			const clock = sinon.useFakeTimers();
+			let isJobCompleted = false;
+
+			const job = new CronJob(
+				'* * * * * *',
+				() => {
+					setTimeout(() => {
+						callback();
+						isJobCompleted = true;
+					}, 500);
+				},
+				null,
+				true,
+				null,
+				null,
+				false,
+				null,
+				false,
+				false // waitForCompletion: false
+			);
+
+			expect(job.isCallbackRunning).toBe(false);
+
+			// First execution
+			clock.tick(1000);
+			expect(isJobCompleted).toBe(false);
+			expect(job.isCallbackRunning).toBe(false);
+
+			job.stop();
+		});
+
+		it('should track multiple running jobs correctly', async () => {
+			const clock = sinon.useFakeTimers();
+			const executionOrder: number[] = [];
+
+			const job = new CronJob(
+				'* * * * * *',
+				async () => {
+					await new Promise<void>(resolve => {
+						setTimeout(() => {
+							callback();
+							resolve();
+						}, 1500);
+					});
+					const jobNumber = executionOrder.length + 1;
+					executionOrder.push(jobNumber);
+				},
+				null,
+				true,
+				null,
+				null,
+				false,
+				null,
+				false,
+				true // waitForCompletion: true
+			);
+
+			await clock.tickAsync(3500);
+
+			expect(executionOrder).toEqual([1, 2]);
```

Comment:
if new jobs are skipped entirely, the job that starts at the first second and is still running at 2 seconds (because it needs to wait 1.5 seconds before finishing) should mean the second job that normally would have started at 2 seconds is skipped, right?

Reply:
Agreed, the array there should just equal `[1]`.

Here's a version of that test that also tracks/asserts the number of callbacks started:

``` ts
it('should track multiple running jobs correctly', async () => {
	const clock = sinon.useFakeTimers();
	const executionOrder: number[] = [];
	let started = 0;

	const job = new CronJob(
		'* * * * * *',
		async () => {
			started++;
			await new Promise<void>(resolve => {
				setTimeout(() => {
					callback();
					resolve();
				}, 1500);
			});
			const jobNumber = executionOrder.length + 1;
			executionOrder.push(jobNumber);
		},
		null,
		true,
		null,
		null,
		false,
		null,
		false,
		true // waitForCompletion: true
	);

	await clock.tickAsync(3500);

	expect(started).toBe(2);
	expect(executionOrder).toEqual([1]);
	job.stop();
});
```

Is the issue on this line?

https://github.com/kelektiv/node-cron/blob/f4e9e021a2217b422d2c4703d05cd8d5308fe8d4/src/job.ts#L210

If the `!` is removed all the tests pass.

Reply:
I think you're right. I'll double check the tests and push a fix